crates/lib/src/execution_data.rs

```diff
@@ -1,120 +1,343 @@
+    fn from_qvm_result_data(
+        result_data: &QvmResultData,
+    ) -> Result<Self, RegisterMatrixConversionError> {
+        Ok(Self(
+            result_data
+                .memory
+                .iter()
+                .map(|(name, register)| {
+                    let register_matrix = match register {
+                        RegisterData::I8(data) => RegisterMatrix::Integer(
+                            Array::from_shape_vec(
+                                (data.len(), data.first().map_or(0, Vec::len)),
+                                data.iter().flatten().copied().map(i64::from).collect(),
+                            )
+                            .map_err(|_| {
+                                RegisterMatrixConversionError::InvalidShape {
+                                    register: name.to_string(),
+                                }
+                            })?,
+                        ),
+                        RegisterData::I16(data) => RegisterMatrix::Integer(
+                            Array::from_shape_vec(
+                                (data.len(), data.first().map_or(0, Vec::len)),
+                                data.iter().flatten().copied().map(i64::from).collect(),
+                            )
+                            .map_err(|_| {
+                                RegisterMatrixConversionError::InvalidShape {
+                                    register: name.to_string(),
+                                }
+                            })?,
+                        ),
+                        RegisterData::F64(data) => RegisterMatrix::Real(
+                            Array::from_shape_vec(
+                                (data.len(), data.first().map_or(0, Vec::len)),
+                                data.iter().flatten().copied().collect(),
+                            )
+                            .map_err(|_| {
+                                RegisterMatrixConversionError::InvalidShape {
+                                    register: name.to_string(),
+                                }
+                            })?,
+                        ),
+                        RegisterData::Complex32(data) => RegisterMatrix::Complex(
+                            Array::from_shape_vec(
+                                (data.len(), data.first().map_or(0, Vec::len)),
+                                data.iter()
+                                    .flatten()
+                                    .copied()
+                                    .map(|c| Complex64::new(c.re.into(), c.im.into()))
+                                    .collect(),
+                            )
+                            .map_err(|_| {
+                                RegisterMatrixConversionError::InvalidShape {
+                                    register: name.to_string(),
+                                }
+                            })?,
+                        ),
+                    };
```

Comment:
Potentially move `map_error` outside of the match statement?

```suggestion
                    let register_matrix = match register {
                        RegisterData::I8(data) => Array::from_shape_vec(
                            (data.len(), data.first().map_or(0, Vec::len)),
                            data.iter().flatten().copied().map(i64::from).collect(),
                        )
                        .map(RegisterMatrix::Integer),
                        RegisterData::I16(data) => Array::from_shape_vec(
                            (data.len(), data.first().map_or(0, Vec::len)),
                            data.iter().flatten().copied().map(i64::from).collect(),
                        )
                        .map(RegisterMatrix::Integer),
                        RegisterData::F64(data) => Array::from_shape_vec(
                            (data.len(), data.first().map_or(0, Vec::len)),
                            data.iter().flatten().copied().collect(),
                        )
                        .map(RegisterMatrix::Real),
                        RegisterData::Complex32(data) => Array::from_shape_vec(
                            (data.len(), data.first().map_or(0, Vec::len)),
                            data.iter()
                                .flatten()
                                .copied()
                                .map(|c| Complex64::new(c.re.into(), c.im.into()))
                                .collect(),
                        )
                        .map(RegisterMatrix::Complex),
                    }
                    .map_err(|_| {
                        RegisterMatrixConversionError::InvalidShape {
                            register: name.to_string(),
                        }
                    })?;
```

I wanted to do this with other common stuff too, but didn't see a clean way to do this
```rs
let arr2 = |data| {
    Array::from_shape_vec(
        (data.len(), data.first().map_or(0, Vec::len)),
        data
    )
};
```

Reply:
Good idea, applied your change.